src/com/session/GoogleResp.java

```diff
@@ -17,45 +20,58 @@
 	private static final long serialVersionUID = 1L;
 
 	public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
-		
-		String accessToken=null;
+
 		String authcode = req.getParameter("code");
-		String email=null;
-		boolean isPresent=false;
-		System.out.println(authcode);
+		String error = req.getParameter("error");
+		GoogleInfo post = null;
+
+		if ("access_denied".equals(error)) {
+			resp.sendRedirect("login?access_denied");
+
+		}
+
 		if (authcode != null) {
 
-			if (authcode.equals("access_denied")) {
-				resp.sendRedirect("login");
+			String accessToken = GoogleService.getAccessToken(authcode);
+
+			if (accessToken == null) {
+				/// login?error=token_fetch_error
+				resp.sendRedirect("login?error=token_fetch_error");
 
-			} else {
-				 accessToken= GoogleService.getAccessToken(authcode);
 			}
 
+			else {
+				post = GoogleService.getUserInfo(accessToken);
+			}
 		}
-		
-		if (accessToken!= null) {
+		if (post != null) {
+			String email = post.getEmail();
+			if (email != null) {
+				boolean isPresent = SessionHelper.isPresent(email);
+
+				if (isPresent) {
+					HttpSession s1 = req.getSession();
+					s1.setAttribute("email", email);
+					resp.sendRedirect("dashboard");
 
-			 email = GoogleService.getUserInfo(accessToken);
+				} else {
+					String name = post.getGiven_name();
+					String password = null;
+					// System.out.println(name);
+					Entity contact = SignUp.addinfo(email, name, password);
+					HttpSession s1 = req.getSession(false);
+					String session = SessionHelper.currentUser(req);
+					if (session == null) {
```

Comment:
session object is checked against null ? the invalidate() method call will throw error